src/main/java/org/jabref/gui/theme/ThemeImpl.java

```diff
@@ -0,0 +1,219 @@
+import org.jabref.preferences.PreferencesService;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * Installs the style file and provides live reloading.
+ * JabRef provides two inbuilt themes and a user customizable one: Light, Dark and Custom. The Light theme is basically
+ * the base.css theme. Every other theme is loaded as an addition to base.css.
+ *
+ * For type Custom, Theme will protect against removal of the CSS file, degrading as gracefully as possible. If the file
+ * becomes unavailable while the application is running, some Scenes that have not yet had the CSS installed may not be
+ * themed. The PreviewViewer, which uses WebEngine, supports data URLs and so generally are not affected by removal
+ * of the file; however Theme will not attempt to URL-encode large style sheets so as to protect
+ * memory usage (see {@link StyleSheetFile#MAX_IN_MEMORY_CSS_LENGTH}.
+ *
+ * @see <a href="https://docs.jabref.org/advanced/custom-themes">Custom themes</a> in the Jabref documentation.
+ */
+public class ThemeImpl implements Theme {
+
+    private static final Logger LOGGER = LoggerFactory.getLogger(ThemeImpl.class);
+
+    private final PreferencesService preferencesService;
+
+    private final FileUpdateMonitor fileUpdateMonitor;
+
+    private final Consumer<Runnable> updateRunner;
+
+    private final StyleSheet baseStyleSheet;
+
+    private final AtomicReference<ThemeState> state = new AtomicReference<>();
+
+    private final Set<Scene> scenes = Collections.newSetFromMap(new WeakHashMap<>());
+    private final Set<WebEngine> webEngines = Collections.newSetFromMap(new WeakHashMap<>());
+
+    private static final class ThemeState {
+
+        private final ThemePreference themePreference;
+
+        private final StyleSheet additionalStylesheet;
+
+        public ThemeState(ThemePreference themePreference) {
+            this.themePreference = themePreference;
+            this.additionalStylesheet = switch (themePreference.getType()) {
+                case LIGHT -> StyleSheetEmpty.EMPTY;
+                case DARK -> StyleSheet.create("Dark.css");
+                case CUSTOM -> StyleSheet.create(themePreference.getName());
+            };
+        }
+    }
+
+    public ThemeImpl(ThemePreference themePreference, PreferencesService preferencesService, FileUpdateMonitor fileUpdateMonitor, Consumer<Runnable> updateRunner) {
+        if (themePreference == null) {
+            throw new IllegalArgumentException("Theme preference required");
+        }
+        if (preferencesService == null) {
+            throw new IllegalArgumentException("Preference service required");
+        }
+        if (fileUpdateMonitor == null) {
+            throw new IllegalArgumentException("File update monitor required");
+        }
+        if (updateRunner == null) {
+            throw new IllegalArgumentException("Update runner required");
+        }
+        this.preferencesService = preferencesService;
```

Comment:
Use `this.preferencesService = Objects.requireNonNull(preferenceService)` instead of the manual null check above.